### PR TITLE
[3.1.1] - 2019-11-12

### PR DESCRIPTION
**Pull request by Worvast**
#### Added
 * Support to CA_MD_TOO_WEAK problems with a new flag "sec_level"
 * New flag documentation
 
#### Changed
 * SSL Socket in Sender creation with custom context
 * Tests Sender certificates generation/files are updated
 
#### Fixed
 * Problems in documentation with API information

### CODE REVIEW

File: devo/sender/data.py [hunk 1]
```diff
@@ -35,16 +35,18 @@ class SenderConfigSSL:
     :param cert:  (str) cert src file
     :param chain:  (str) chain src file
     :param pkcs:  (dict) (path: pfx src file, password: of cert)
+    :param sec_level: (int) default None. If certs are too weak you can change
+    this param to work with it
```
Comment (aarteagamel) [inline devo/sender/data.py]: What are the valid values here? maybe adding a link to an url can be helpful.

File: devo/sender/data.py [hunk 2]
```diff
@@ -188,12 +190,22 @@ def __connect_ssl(self):
...
-                        ca_certs=self._sender_config.chain,
-                        cert_reqs=ssl.CERT_REQUIRED)
+
+                    context = ssl.create_default_context(
+                        cafile=self._sender_config.chain)
+
+                    if self._sender_config.sec_level is not None:
+                        context.set_ciphers(
+                            "DEFAULT@SECLEVEL={!s}"
+                            .format(self._sender_config.sec_level))
```
Comment (aarteagamel) [inline devo/sender/data.py]: Do we have a logger here? I think that is a good idea to log a WARNING message when the sec_level is different than the default

File: docs/sender.md [hunk 3]
```diff
@@ -53,7 +53,8 @@ Class SenderConfigSSL accept various types of certificates, you has:
 + key **(_str_)**: key src file
 + cert **(_str_)**: cert src file
 + chain **(_str_)**: chain src file
-+ pkcs **(_dict_)**: (path: pfx src file, password: of cert)
++ pkcs **(_dict_)**: (path: pfx src file, password: of cert
```
Comment (aarteagamel) [inline docs/sender.md]: Why did you remove the end parenthesis?

File: docs/sender.md [hunk 4]
```diff
@@ -209,6 +210,20 @@ compression_level is an integer from 0 to 9 or -1 controlling the level of compr
  (1, each time, if not zipped, 0..X if zipped)
  
  
+### CA_MD_TOO_WEAK
+Or CA signature digest algorithm too weak its a error with news versions of openssl>=1.1.0
+If you have problem with your certificates of Devo and devo-sdk you with this error you can add flag `sec_level=0`
```
Comment (aarteagamel) [inline docs/sender.md]: Idem as function doc. If we can add some info regarding valid values will be helpful

File: devo/sender/data.py (hunk 1 above)
Comment (Worvast) [reply devo/sender/data.py]: Updated in documentation

File: docs/sender.md (hunk 3 above)
Comment (Worvast) [reply docs/sender.md]: Because error. Fixed.

File: docs/sender.md (hunk 4 above)
Comment (Worvast) [reply docs/sender.md]: Added more info with all levels.

File: devo/sender/data.py (hunk 2 above)
Comment (Worvast) [reply devo/sender/data.py]: Added on warning message
